---
title: "Summary Re-write of Sierra's Replicate Analysis Using Tidyverse"
date: '`r Sys.Date()`'
output: 
    bookdown::html_document2:
        toc: true
        toc_float: true
        df_print: paged
        code_folding: hide
---

```{r setup}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse); packageVersion("tidyverse")
library(yaml); packageVersion("yaml")
library(here); packageVersion("here")
library(stringr); packageVersion("stringr")
```

# Summary
Determine sources of variability for the CHANGE-Seq assay.

# Approach

1. Samples are run through the CHANGE-Seq wet lab procedure and the library prepped smples are split between NIST & St. Jude. 
2. Samples are pooled and sequenced. 
3. Resulting FASTQ files are handed off to me and run through the CHANGE-Seq program. 
4. Each samples produces an "identified_matched.txt" file which is what is used and described in this document. 

**Downstream Bioinformatic Analysis:**

1. Loading CHANGE-Seq pipeline output (namely - "identified_matched.txt" files)
2. Annotating CHANGE-Seq data frame with sample pair/ replicate information
3. Summary/ Statistical Analysis  
    1. Summary Table  
    2. StJude - NIST Comparison

# Data Overview
## Samples A-F Spring 2020

**Letter ID** | **Target Site**
--------------|-----------------
A             | CCR5_site_3
B             | CCR5_site_8
C             | TRAC_site_1
D             | CXCR4_site_7
E             | CTLA4_site_9
F             | AAVS1_site_14

**Experimental Factors**
N: NIST
S: St. Jude

**Acronym** | **Wet Lab** | **Sequencing Lab** | **Bioinformatics**
------------|-------------|--------------------|--------------------
NNN         | NIST        | NIST               | NIST
NSN         | NIST        | St. Jude           | NIST
NSS         | NIST        | St. Jude           | St. Jude

__NOTE__ SSN: refers to the CHANGE-Seq manuscript/publication samples. See below for pre-run steps
**Pending review by Cicera as samples do not seem to be matched depth-wise when compared. 

__NOTE__ NNN2: NNN same library prep sequenced on an expired MiSeq kit (details below) to test before running new prepped AEF replicate samples. When these were of adequate quality, we incorporated them into this analysis. 

*St Jude has not yet re-run our sequencing of the original A-F samples. (NNS)*

1. Loading CHANGE-Seq pipeline output (namely - "identified_matched.txt" files)
2. Annotating CHANGE-Seq data frame with sample pair/ replicate information
3. Summary/ Statistical Analysis  
    1. Summary Table  
    2. Interlaboratory Comparisons

## Data Pre-Run Steps

**Reference Genome**
Cicera:
> You can download hg38 sequence from: https://hgdownload.soe.ucsc.edu/goldenPath/hg38/chromosomes/ and concatenate only the main chromosome sequences as hg38_chroms_only.fa

Downloaded all files with this naming scheme: "chrom1.fa.gz" + X & Y. 

Concatenated files via:

```{r, engine = 'bash', eval = FALSE}
cd ~/Desktop/reference_genome
cat *.fa > ~/Desktop/reference_genome/combined_hg38_chroms.fa
```

**Trimmed NIST-sequenced files due to extra cycling with our kit via:**

```{r, engine = 'bash', eval = FALSE}
seqtk trimfq -e 150 \
A_S1_L001_R1_001.fastq.gz > A_S1_L001_R1_001_trimmed.fq
```


**SSN files:** 
Cicera:
> Those samples were sequenced on a MiSeq V3 kit. For comparison of replicates, you will need to sample the gz files to the same sequence depth (between the replicates), and ideally, once you get the results, you would merge the tables and set a cutoff (read counts), because the low frequency sites are very likely to occur in one replicate but not in other. That's what we use for the comparisons in figure 1. (edited) 

>you can use this command to check how many reads in each file:

```{r, engine = 'bash', eval = FALSE}
parallel "echo {} && gunzip -c {} | wc -l | awk '{d=\$1; print d/4;}'" ::: *.gz
```

>Then, you can use this one to generate new files with the number of reads that you want:

```{r, engine = 'bash', eval = FALSE}
seqtk sample -s100 read1.fq 1000 > sub1.fq
```

where 1000 is the number you want to sample to. **NOTE** She rerefences merging the tables but I prefer to maintain them separately and compare them as replicates. 

### NN

**Sample**       | **File Name**                             | **File Size** | **Depth**
-----------------|-------------------------------------------|---------------|------------
A, read 1        | A_S1_L001_R1_001_trimmed.fq               | 956.9 MB      | 2631758
A, read 2        | A_S1_L001_R2_001_trimmed.fq               | 874.8 MB      | 2631758
B, read 1        | B_S2_L001_R1_001_trimmed.fq               | 647.9 MB      | 1780610
B, read 2        | B_S2_L001_R2_001_trimmed.fq               | 623.7 MB      | 1780610
C, read 1        | C_S3_L001_R1_001_trimmed.fq               | 743.2 MB      | 2049156
C, read 2        | C_S3_L001_R2_001_trimmed.fq               | 706.6 MB      | 2049156
D, read 1        | D_S4_L001_R1_001_trimmed.fq               | 935.6 MB      | 2586872
D, read 2        | D_S4_L001_R2_001_trimmed.fq               | 739.4 MB      | 2586872
E, read 1        | E_S5_L001_R1_001_trimmed.fq               | 741.1 MB      | 2056395
E, read 2        | E_S5_L001_R2_001_trimmed.fq               | 701.1 MB      | 2056395
F, read 1        | F_S6_L001_R1_001_trimmed.fq               | 845.6 MB      | 2339851
F, read 2        | F_S6_L001_R2_001_trimmed.fq               | 817.3 MB      | 2339851
Neg Ctrl, read 1 | neg-neg-control_S7_L001_R1_001_trimmed.fq | 858.1 MB      | 2341975
Neg Ctrl, read 2 | neg-neg-control_S7_L001_R2_001_trimmed.fq | 803 MB        | 2341975

### NS

**Sample**       | **File Name**                             | **File Size** | **Depth**
-----------------|-------------------------------------------|---------------|------------
A, read 1        | CRL1551_NIST_S1_R1_001.fastq              | 956.9 MB      | 2829840
A, read 2        | CRL1551_NIST_S1_R2_001.fastq              | 874.8 MB      | 2829840
B, read 1        | CRL1552_NIST_S2_R1_001.fastq              | 647.9 MB      | 2453359
B, read 2        | CRL1552_NIST_S2_R2_001.fastq              | 623.7 MB      | 2453359
C, read 1        | CRL1553_NIST_S3_R1_001.fastq              | 743.2 MB      | 2084289
C, read 2        | CRL1553_NIST_S3_R2_001.fastq              | 706.6 MB      | 2084289   
D, read 1        | CRL1554_NIST_S4_R1_001.fastq              | 935.6 MB      | 2840809
D, read 2        | CRL1554_NIST_S4_R2_001.fastq              | 739.4 MB      | 2840809
E, read 1        | CRL1555_NIST_S5_R1_001.fastq              | 741.1 MB      | 2339453
E, read 2        | CRL1555_NIST_S5_R2_001.fastq              | 701.1 MB      | 2339453
F, read 1        | CRL1556_NIST_S6_R1_001.fastq              | 845.6 MB      | 2416305
F, read 2        | CRL1556_NIST_S6_R2_001.fastq              | 817.3 MB      | 2416305
Neg Ctrl, read 1 | neg-neg-CRL1557_NIST_S7_R1_001.fastq      | 858.1 MB      | 2604617
Neg Ctrl, read 2 | neg-neg-CRL1557_NIST_S7_R2_001.fastq      | 803 MB        | 2604617

### Notes Sample Sequencing

![Sequencing Information for A-F Samples at NIST](/Users/sdm8/Documents/Projects/CHANGEseq/analysis/A-F/2020-08-20/Screen Shot 2020-08-20 at 12.16.05 PM.png)
![Original sequencing of A-F at NIST - Spring 2020](/Users/sdm8/Documents/Projects/CHANGEseq/analysis/A-F/2020-08-20/Screen Shot 2020-08-20 at 12.16.31 PM.png)
![Re-sequencing of same library prep as spring2020 A-F samples at NIST. Run on expired MiSeq kit to test machine before running replicates of new prep A, E, F](/Users/sdm8/Documents/Projects/CHANGEseq/analysis/A-F/2020-08-20/Screen Shot 2020-08-20 at 12.16.38 PM.png)

To account for extra sequencing cycles, samples sequenced at NIST were trimmed via: 
```{r, engine = 'bash', eval = FALSE}
seqtk trimfq -e 150
```

Both locations pooled samples according to St. Jude's qPCR quantification values. 

Samples run with CHANGE-Seq v1.2.7

*St Jude has not yet re-run our sequencing of the original A-F samples. (NNS)*

**NOTE About SSN Files**

I sampled the files per Cicera's instructions and checked to make sure the number of sequences matched those of St. Jude's A-F sequencing of our samples. However, when I started to plot them with our other samples they drastically outweighed ours and made it very difficult to see our samples on the graphs. I will include two examples here and we can discuss how we want to proceed. We may just need to compare within sequencing depths and not try to down-sample. 

![Total read count of all samples showing how the SSN replicates even when down-sampling drowns out our samples](/Users/sdm8/Desktop/changeseq-analysis/scratch/Screen Shot 2020-08-18 at 11.56.31 AM.png)

![Total on-target read count with the down-sampled manuscript SSN replicates showing that it makes it very difficult to see our other samples when plotted together](/Users/sdm8/Desktop/changeseq-analysis/scratch/Screen Shot 2020-08-18 at 11.56.36 AM.png)

# Analysis
## Loading and Munging data 

Loading CHANGE-Seq output txt files into a single data frame
```{r message = FALSE}
## Using message = FALSE to avoid printing read_tsv message about column specifications

## List of changeseq output files
changeseq_lst <- list.files(path = here("data-raw"), 
                           pattern = "identified_matched.txt",
                           include.dirs = TRUE, full.names = TRUE, 
                           recursive = TRUE) %>% 
    set_names(basename(.))

changeseq_df <- changeseq_lst %>%
    map_dfr(read_tsv, .id = "changeseq_run")
```


```{r}
unique(changeseq_df$changeseq_run)
```


__NOTE__ - can modify data frame to add lab name, e.g. StJude or NIST, either using file paths or separate data frame.
Annotating using file names
```{r}
## This code is specific to the input file name format
changeseq_anno_df <- changeseq_df %>% 
    ## adding changeseq_run col, removing ending portion of string **NOTE THAT THIS WILL NEED CHANGED WHEN LOOKING AT OTHER DNA LINES**
    mutate(changeseq_run = str_remove(changeseq_run, "_GM24385_identified_matched.txt"),
           sample = "GM24385",
           ## target site name begins 5 characters into string of changeseq_run
          target_site = str_sub(changeseq_run, 5),
          ## further edit target site col if contains the St Jude file prefixes of CRLXXXX
           target_site = str_remove(target_site,  "CRL[:digit:]*_"),
            target_site = str_remove(target_site, "_sampled_"),
          ## lab col will contain the first 3 letter acronyms
          lab = substr(changeseq_run, 1,4),
             lab = str_remove(lab, "_")) %>%
     ## possible identifiers: NNN, NSN, NSS
     ## first = wet lab, second = sequence, third = bioinformatics
    ## take the 3 letters in lab col and split into 3 additional cols 
    separate(lab, into = c("wet_lab","seq_lab","bioinf"), 
           sep = c(1,2,3),
           remove = FALSE) %>%
  ## change the col contents from just letters to what they indicate
  mutate(wet_lab = if_else(wet_lab == "N", "NIST", "StJude"),
         seq_lab = if_else(seq_lab == "N", "NIST", "StJude"),
         bioinf = if_else(bioinf == "N", "NIST", "StJude"))

changeseq_anno_df ## Samantha may want this one written out
```

__NOTE__ - The lab, target_site, and/ or changeseq_run variables can be used as grouping variables to calculate within dataset summarise.

## Calculating Summary Metrics

**Total Read Counts per Sample**
```{r}
total_count_df <- changeseq_anno_df %>% 
    group_by(changeseq_run, lab, target_site) %>% 
    summarise(total_read_count = sum(Nuclease_Read_Count))
```

**Total on-target read count**
```{r}
target_count_df <- changeseq_anno_df %>% 
    filter(Site_Substitution_Number == 0) %>% 
    rename(on_target_count = Nuclease_Read_Count) %>% 
    select(changeseq_run, lab, target_site, on_target_count)
```

**Percent on-target read count (on-target read count/total read count)**
```{r}
on_off_target_df <- left_join(total_count_df, target_count_df) %>% 
    mutate(on_target_rate = on_target_count/total_read_count)
```

**Total off-target read count**
**Percent off-target read count (off-target read count/total read count)**
```{r}
summary_df <- changeseq_anno_df %>% 
    ## replace NAs in Site_Substitution_Number col
    mutate(Site_Substitution_Number = if_else(is.na(Site_Substitution_Number), -1, Site_Substitution_Number), 
           ## assign read count to on-target col of on-target row per sample
            on_target = if_else(Site_Substitution_Number == 0, Nuclease_Read_Count, 0),
           ## assign off-target read counts in non-on-target rows
            off_target = if_else(on_target == 0, Nuclease_Read_Count, 0),
            ## seqs with subs only = seq in Site_Sequence col only
            Sub_Only = if_else(is.na(Site_Sequence), FALSE, TRUE),
            ## seqs with indels only = seq in Site_Sequence_Gaps_Allowed col only
            Indel_Only = if_else(is.na(Site_Sequence_Gaps_Allowed), TRUE, FALSE),
            ## seqs with two possible reps = seq in both cols
            Sub_or_Indel = if_else(is.na(Sub_Only || Indel_Only), FALSE, TRUE))
```
  
**Number of sequences with substitutions only**
**Number sequences with indels only**
**Number of sequences with two possible representations (substitutions or indels)**
```{r}
## create df to plot off-target ratio
off_target_ratio <- summary_df %>%  
    group_by(changeseq_run, lab, target_site) %>% 
    summarise(total_count = sum(Nuclease_Read_Count),
              total_on_target = sum(on_target),
              total_off_target = sum(off_target),
              off_target_ratio = total_off_target/total_count)
## create df to plot alignment of off-target seq resolution details
off_target_seq_align_df <- summary_df%>%    
    group_by(changeseq_run, lab, target_site) %>%
    summarise(total_sub_only_seqs = sum(Sub_Only),
              total_indels_only = sum(Indel_Only),
              total_sub_or_indel = sum(Sub_or_Indel))
```

__NOTE__ CONSIDERING SEQUENCES THAT HAVE SUBSTITUTIONS ONLY...
**On-target to top 25% reads (on-target read count/top 25% off-target read count)**
```{r}
## create df of all off-target rows sorted by decreasing read count value
off_target_df <- summary_df %>%
    mutate(Substitution_Only = if_else(is.na(Site_Sequence), FALSE, TRUE),
           on_target = if_else(Site_Substitution_Number == 0, Nuclease_Read_Count, 0)) %>% 
    filter(Substitution_Only == TRUE) %>%
        arrange(-Nuclease_Read_Count)

## take the top 25% of the off-target df - based on read count value
top_quarter_off_target_df <- off_target_df %>%
    ## Use top_frac to get the top 25% - use group_by to capture within run or site analysis
      top_frac(0.25, wt = Nuclease_Read_Count) %>%
  group_by(changeseq_run, lab, target_site) %>% 
  summarise(top_quarter_offtarget_readcount = sum(Nuclease_Read_Count))
```

**On-target to most prevalent off-target read (on-target read count/off-target read with largest read count)**
```{r}
## create df to plot ratio between on-target read count and most prevalent off-target read count per sample
on_top_off_df <- summary_df %>% 
    mutate(Site_Substitution_Number = if_else(is.na(Site_Substitution_Number), -1, Site_Substitution_Number), 
            on_target = if_else(Site_Substitution_Number == 0, Nuclease_Read_Count, 0),
           off_target = if_else(on_target == 0, Nuclease_Read_Count, 0)) %>% 
    group_by(changeseq_run, lab, target_site) %>% 
    summarise(on_target_count = max(on_target),
              off_target_max = max(off_target)) %>% 
    mutate(on_off_max_ratio = on_target_count/ off_target_max)
```

**Read count for seqs w/ 1 substitution from target sequence**
**Read count for seqs w/ 2 substitutions from target sequence**
**Read count for seqs w/ 3 substitutions from target sequence**
**Read count for seqs w/ 4 substitutions from target sequence**
**Read count for seqs w/ 5 substitutions from target sequence**
**Read count for seqs w/ 6 substitutions from target sequence**
```{r}
## create df to plot the binned read counts of off-target substitution numbers
substitution_readcount_df <- summary_df %>%
    mutate(Substitution_Only = if_else(is.na(Site_Sequence), FALSE, TRUE)) %>% 
    filter(Substitution_Only == TRUE,
           Site_Substitution_Number > 0) %>% 
    group_by(changeseq_run, lab, target_site, Site_Substitution_Number) %>% 
    summarise(substitution_read_count = sum(Nuclease_Read_Count))
```

## Table

```{r}
## Making substitution count data frame wide for join
sub_counts_wide_df <- substitution_readcount_df %>% 
  mutate(Site_Substitution_Number = paste("sub",Site_Substitution_Number, "count", 
                                          sep = "_")) %>%
  pivot_wider(names_from = Site_Substitution_Number, 
              values_from = substitution_read_count, values_fill = 0)

## Combining summary metric data frames
summary_table <- left_join(off_target_ratio, off_target_seq_align_df) %>% 
  left_join(top_quarter_off_target_df) %>% 
  left_join(on_top_off_df) %>% 
  left_join(sub_counts_wide_df)

## write out summary table
## Note - do not need to use paste with here, here converts the vector of character strings to a file path
write_tsv(summary_table, path = here("data", "summary_stats_table.tsv"), 
          na = "NA", append = FALSE, col_names = TRUE, quote_escape = "double")
```


```{r}
## Potentially use the DT or other packages for visually pleasing table presentation
summary_table %>% DT::datatable()
```

**Summary Stats Table Column Headers**

 [1] **"changeseq_run":** lab identifiers, sample letter ID, target site                  
 [2] **"lab":** 3-letter identifiers for wet lab, sequencing lab, and bioinformatics lab; **N**IST, **S**t. Jude                           
 [3] **"target_site":** target site                   
 [4] **"total_count":** total read count for that sample (both on and off-target)                   
 [5] **"total_on_target":** total read count for the on-target sequence for that sample                
 [6] **"total_off_target":** total read count for the off-target sequences for that sample                
 [7] **"off_target_ratio":** total on-target read count/total off-target read count               
 [8] **"total_sub_only_seqs":** total read count for the off-target sequences that are resolved in alignment with nucleotide substitutions only  
 [9] **"total_indels_only":** total read count for the off-target sequences that are resolved in alignment with gaps (indels) only              
 [10] **"total_sub_or_indel":** total read count for the off-target sequences that are resolved in alignment with substitutions or gaps             
 [11] **"top_quarter_offtarget_readcount":** the total read count for the top quarter of off-target reads (sorted by decreasing read count) 
 [12] **"off_target_max":** the read count for the most prevalent off-target sequence (off-target sequence with largest read count)                 
[13] **"on_off_max_ratio":** on-target read count/read count for most prevalent off-target sequence                 
[14] **sub_X_count:** the unique substitution number found across the samples with the total number of occurrences of that read number in the off-target coordinates across each sample

## Plots
### Total Read Counts & On-Target Read Counts per Sample

For each target site, each lab's total read count is plotted. This gives an idea how much data is generated as a whole for each target site. For each target site, each lab's total on-target read count is plotted. 


```{r}
ggplot(total_count_df) + 
    geom_bar(aes(x = target_site, y = total_read_count, fill = lab), 
             position = "dodge", stat = "identity") +
    theme_bw() + 
    theme(axis.text.x = element_text(angle = 90)) + 
    labs(x = "Editing Target Site", y = "Total Read Count", 
         fill = "Lab") 

ggplot(target_count_df) + 
    geom_bar(aes(x = target_site, y = on_target_count, fill = lab), 
             position = "dodge", stat = "identity") +
    theme_bw() + 
    theme(axis.text.x = element_text(angle = 90)) + 
    labs(x = "Editing Target Site", y = "Total On-Target Read Count", 
         fill = "Lab")
```


Sample F for NSS has a drastically lower read count. 
This was found to be a copy/paste error in their manifest file. 
They were pointing to sample E's read fastq files. 

### On-Target Rate

For each target site, each lab's on target rate (total on-target read count / total read count) is plotted. 

``````{r}
ggplot(on_off_target_df) + 
    geom_point(aes(x = target_site, y = on_target_rate, fill = lab), 
               shape = 21, alpha = 0.5) + 
    theme_bw() +
    theme(axis.text.x = element_text(angle = 90)) +
        labs(x = "Editing Target Site", y = "On Target Rate", 
         fill = "Lab")
```

On-target rates for sample A, E, and F are nearly matched and variance for samples B, C, and D is observed. NIST-sequenced samples are slightly higher.

### Off-Target Rate

For each target site, each lab's off-target rate (off-target read count / total read count) is plotted.

``````{r}
ggplot(off_target_ratio) + 
    geom_point(aes(x = target_site, y = off_target_ratio, fill = lab), 
               shape = 21, alpha = 0.5) + 
    theme_bw() +
    theme(axis.text.x = element_text(angle = 90)) +
        labs(x = "Editing Target Site", y = "Off Target Rate", 
         fill = "Lab")
```

The rates of the off-target sequences are generally all higher than that of the on-target rates. Like the on-target rates, samples A, E, and F are similar in value. Samples B, C, and D have variance with the inverse of the above - the off-target rates are higher for the St. Jude-sequenced samples than that of the NIST-sequenced samples. 


### Off-target Sequence Alignment: Sub/Gap

For each target site, the number of off-target sequences that are resolved by substitutions only, indels (gaps) only, or substitutions *or* gaps are plotted for each lab. 

```{r}
off_target_seq_align_df %>% 
    ## First make the table long
    gather(key = "metric", value = "value", -changeseq_run, -lab, -target_site)
```

```{r}
off_target_seq_align_df %>% 
    ## First make the table long
    gather(key = "metric", value = "value", -changeseq_run, -lab, -target_site) %>% 
    ggplot() +
    geom_bar(aes(x = metric, 
                 y = value,
                 fill = lab),
             position = "dodge", 
             stat = "identity"
             ) +
    ## Split plot by summary metric
    facet_wrap(~target_site, scales = "free_y") +
    theme_bw() +
    theme(axis.text.x = element_text(angle = -45, hjust = 0)) +
  labs(x = "Mismatch Type", y = "Number of Sequences", fill = "Lab")
```

The off-target sequences are aligned to the target sequence. These are binned in the following manner: in order to align properly to the target sequence, the off-target sequence could be resolve with substitutions (and assigned a substitution number), gaps (insertions or deletions) or by both possibilities. Here, I have binned them accordingly. As before, no differences are observed between the samples sequenced at St. Jude but run through the CHANGE-Seq bioinformatics pipeline by me and St. Jude, however, variance is observed between the samples sequenced at NIST vs. St. Jude. For samples A and B, NIST-sequenced off-target reads are overall in lower numbers as compared to St. Jude.


### Top 25% off-target read counts

For each target site, weighted by read count, the top 25% off-target sequence read counts for each lab are plotted. 

```{r}
ggplot(top_quarter_off_target_df) + 
    geom_bar(aes(x = target_site, y = top_quarter_offtarget_readcount, fill = lab), 
             position = "dodge", stat = "identity") +
    theme_bw() + 
    theme(axis.text.x = element_text(angle = 90)) + 
    labs(x = "Total Read Count of Top 25% Off-Targets", y = "Editing Target Site", 
         fill = "Lab")
```

### Ratio of on-target read counts to the off-target sequence with the highest read count

The on-target read count divided by the most prevalent off-target sequence (the off-target sequence with the largest read count) for each sample is plotted for each lab. 

``````{r}
ggplot(on_top_off_df) + 
    geom_point(aes(x = target_site, y = on_off_max_ratio, fill = lab), 
               shape = 21, alpha = 0.5) + 
    theme_bw() +
    theme(axis.text.x = element_text(angle = 90)) +
        labs(x = "Editing Target Site", y = "On-Target Rate to Most Prevalent Off-Target", 
         fill = "Lab")
```

As seen with the on-target rate considering total read counts, NIST has higher rates than St. Jude-sequenced samples. 

### Read counts per subsitution number for each sample

For each substitution number (or for indels only or subs/indels which are grouped separately), the number of off-target sequences that are classified as such as such are plotted for each target sample and lab. 

__NOTE__ I currently have these plotting read counts but we may want to either replace this with or add another graph that shows occurrences for each sub number.

```{r}
indel_df <- off_target_seq_align_df %>% 
  select(-total_sub_only_seqs) %>% 
  pivot_longer(names_to = "diff_cat", values_to = "read_count", 
               cols = c("total_indels_only","total_sub_or_indel"))

diff_cat_read_count_df <- substitution_readcount_df %>% 
  mutate(diff_cat = str_c("Substitution #", Site_Substitution_Number, sep = "")) %>% 
  select(-Site_Substitution_Number) %>% 
  rename(read_count = substitution_read_count) %>% 
  bind_rows(indel_df)

ggplot(diff_cat_read_count_df) + 
    geom_bar(aes(x = diff_cat, 
                 y = read_count, fill = lab), 
             position = "dodge", stat = "identity") +
  facet_wrap(~target_site) + 
    theme_bw() + 
    theme(axis.text.x = element_text(angle = 90)) + 
    labs(x = "Difference Type", y = "Read Count", 
         fill = "Lab")
```


## Venn Bar Plots

As opposed to the typical venn diagrames, these show relative abundance and give appropriate scaling to visualize values. 

__NOTE__ NATE - can we add the values on the graph??

```{r}
off_target_scatter_df <- off_target_df %>% 
  select(Chromosome, Start, End, Nuclease_Read_Count, target_site, lab) %>%
  filter(lab == "NSS")

off_target_lab_set <- off_target_df %>%
  select(`Genomic Coordinate`, target_site, lab, Nuclease_Read_Count) %>% 
  group_by(`Genomic Coordinate`, target_site) %>% 
  summarise(lab_set = str_c(sort(lab), collapse = "-"))

venn_df <- off_target_df %>%
  select(`Genomic Coordinate`, target_site, lab, Nuclease_Read_Count) %>%
  pivot_wider(names_from = "lab", values_from = "Nuclease_Read_Count", values_fill = NA) %>%
  left_join(off_target_lab_set)

ggplot(data = venn_df) + 
  geom_bar(aes(x = target_site, fill = lab_set), position = "fill") + 
  coord_flip() + 
  theme_bw() + 
  labs(x = "Target Site", 
       y = "Proportion of Coordinates", 
       fill = "Lab Combinations")
```

## Read Count Replicate v Replicate Scatter Plots
### Overlapping Off-Target Coordinates

To mimic the plots of Figure1, i in the CHANGE-Seq manuscript, common genomic coordinates of off-targets for each sample replicates are plotted in log scale. 

```{r}
## NSN vs. NSS: Concordant Off-Target Coordinates
nsn_v_nss <- off_target_scatter_df %>% 
  pivot_wider(names_from = "lab", values_from = "Nuclease_Read_Count", values_fill = 0) %>% 
   filter(NSS != 0, NSN != 0) %>%
  ggplot(aes(x = NSN, y = NSS)) + 
   geom_point(alpha = 0.25) + 
   ggpubr::stat_cor(aes(label = ..rr.label..), method = "pearson") + 
  scale_x_log10() + 
  scale_y_log10() + 
  facet_wrap(~target_site, ncol = 1) + 
  theme_bw()

## NSN vs. NNN: Concordant Off-Target Coordinates
nsn_v_nnn <- off_target_scatter_df %>% 
  pivot_wider(names_from = "lab", values_from = "Nuclease_Read_Count", values_fill = 0) %>% 
   filter(NSN != 0, NNN != 0) %>%
  ggplot(aes(x = NSN, y = NNN)) + 
   geom_point(alpha = 0.25) + 
   ggpubr::stat_cor(aes(label = ..rr.label..), method = "pearson") + 
  scale_x_log10() + 
  scale_y_log10() + 
  facet_wrap(~target_site, ncol = 1) + 
  theme_bw()

## NSS vs. NNN: Concordant Off-Target Coordinates
nss_v_nnn <- off_target_scatter_df %>% 
  pivot_wider(names_from = "lab", values_from = "Nuclease_Read_Count", values_fill = 0) %>% 
   filter(NSS != 0, NNN != 0) %>%
  ggplot(aes(x = NSS, y = NNN)) + 
   geom_point(alpha = 0.25) + 
   ggpubr::stat_cor(aes(label = ..rr.label..), method = "pearson") + 
  scale_x_log10() + 
  scale_y_log10() + 
  facet_wrap(~target_site, ncol = 1) + 
  theme_bw()
```

```{r fig.height= 6}
ggpubr::ggarrange(nsn_v_nss, nsn_v_nnn, nss_v_nnn, ncol = 3)
```

## Unique Read Count Occurrence Plots
The read count values for the off-target coordinates unique to replicate 1, replicate 2, and the intersection of the replicates was extracted. Unique read count values from all 3 data were then used to tally each occurrence across the unique genomic coordinates for replicate 1, replicate 2. and the intersection of the replicates

**Our traditional way of viewing**

```{r fig.height = 4}
nnn_nsn_lab_set <- off_target_df %>%
  select(`Genomic Coordinate`, target_site, lab, Nuclease_Read_Count) %>% 
  group_by(`Genomic Coordinate`, target_site) %>% 
  filter(lab != "NSS") %>% 
  summarise(lab_set = str_c(sort(lab), collapse = "-"))

rep_count_scatter_df <- off_target_df %>%
  select(`Genomic Coordinate`, target_site, lab, Nuclease_Read_Count) %>% 
  filter(lab != "NSS") %>% 
  right_join(nnn_nsn_lab_set) %>% 
  count(target_site, lab_set, Nuclease_Read_Count, name = "Occurence")
  
ggplot(rep_count_scatter_df) + 
  geom_point(aes(x = Nuclease_Read_Count, y = Occurence, fill = lab_set), shape = 21) + 
  scale_x_log10() + 
  facet_wrap(~target_site, scales = "free_y", ncol = 1) +
  theme_bw()
```


**Density view**

```{r}
nnn_nsn_lab_set <- off_target_df %>%
  select(`Genomic Coordinate`, target_site, lab, Nuclease_Read_Count) %>% 
  group_by(`Genomic Coordinate`, target_site) %>% 
  filter(lab != "NSS") %>% 
  summarise(lab_set = str_c(sort(lab), collapse = "-"))

rep_count_scatter_df <- off_target_df %>%
  select(`Genomic Coordinate`, target_site, lab, Nuclease_Read_Count) %>% 
  filter(lab != "NSS") %>% 
  right_join(nnn_nsn_lab_set)

ggplot(rep_count_scatter_df) + 
  geom_histogram(aes(x = Nuclease_Read_Count, fill = lab_set)) + 
  scale_x_log10() + 
  facet_wrap(~target_site, scales = "free_y", ncol = 1) +
  theme_bw()

```

**This is plotting the proportion of sites (occurrence / total read count) as opposed to just occurrence number as an attempt to normalize**

```{r}
rep_count_scatter_df %>% 
  mutate(lab_set = factor(lab_set, level = c("NNN-NSN", "NNN","NSN"))) %>% 
ggplot() + 
  geom_histogram(aes(x = Nuclease_Read_Count, fill = lab_set), position = "fill") + 
  scale_x_log10() + 
  facet_wrap(~target_site, scales = "free_y", ncol = 1) +
  theme_bw() +
  scale_fill_brewer(type = "qual", palette = 2) +
  labs(x = "Read Count", y = "Proportion of Sites")

```

# Session Information
## System Information
```{r}
sessioninfo::platform_info()
```

## Package Versions
```{r}
sessioninfo::package_info() %>% 
    filter(attached = TRUE) %>% 
    select(package, loadedversion, date, source) %>%
    knitr::kable(booktabs = TRUE, row.names = FALSE)
```
