---
title: "Re-write of Sierra's Replicate Analysis Using Tidyverse"
date: '`r Sys.Date()`'
output: 
    bookdown::html_document2:
        toc: true
        toc_float: true
        df_print: paged
        code_folding: hide
---

# Questions for rewrite analysis
1. Input data files, specifically list of `identified_matched.txt` files and replicate information.

```{r setup}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse); packageVersion("tidyverse")
library(yaml); packageVersion("yaml")
library(here); packageVersion("here")
library(stringr); packageVersion("stringr")
```

# Summary
Rewriting Sierra's initial replicate analysis for evaluating within and between lab changeseq method performance.

# Approach
<!-- Brief description of individual analysis steps -->

**Experimental Factors**
N: NIST
S: St. Jude

**Acronym** | **Wet Lab** | **Sequencing Lab** | **Bioinformatics**
------------|-------------|--------------------|--------------------
NNN         | NIST        | NIST               | NIST
NSN         | NIST        | St. Jude           | NIST
NSS         | NIST        | St. Jude           | St. Jude

*St Jude has not yet re-run our sequencing of the original A-F samples. (NNS)*

1. Loading CHANGE-Seq pipeline output (namely - "identified_matched.txt" files)
2. Annotating CHANGE-Seq data frame with sample pair/ replicate information
3. Summary/ Statistical Analysis  
    1. Summary Table  
    2. StJude - NIST Comparison

# Analysis
## Loading and Munging data (1 and 2 above)

Loading changeseq output tsv files into a single data frame
```{r message = FALSE}
## Using message = FALSE to avoid printing read_tsv message about column specifications

## List of changeseq output files
changeseq_lst <- list.files(path = here("data-raw"), 
                           pattern = "identified_matched.txt",
                           include.dirs = TRUE, full.names = TRUE, 
                           recursive = TRUE) %>% 
    set_names(basename(.))

changeseq_df <- changeseq_lst %>%
    map_dfr(read_tsv, .id = "changeseq_run")
```


```{r}
unique(changeseq_df$changeseq_run)
```


__NOTE__ - can modify data frame to add lab name, e.g. StJude or NIST, either using file paths or separate data frame.
Annotating using file names
```{r}
## This code is specific to the input file name format
changeseq_anno_df <- changeseq_df %>% 
    ## adding changeseq_run col, removing ending portion of string **NOTE THAT THIS WILL NEED CHANGED WHEN LOOKING AT OTHER DNA LINES**
    mutate(changeseq_run = str_remove(changeseq_run, "_GM24385_identified_matched.txt"),
           sample = "GM24385",
           ## target site name begins 5 characters into string of changeseq_run
          target_site = str_sub(changeseq_run, 5),
          ## further edit target site col if contains the St Jude file prefixes of CRLXXXX
           target_site = str_remove(target_site,  "CRL[:digit:]*_"),
          ## lab col will contain the first 3 letter acronyms
           lab = substr(changeseq_run, 1,3)) %>% 
     ## possible identifiers: NNN, NSN, NSS
     ## first = wet lab, second = sequence, third = bioinformatics
    ## take the 3 letters in lab col and split into 3 additional cols 
    separate(lab, into = c("wet_lab","seq_lab","bioinf"), 
           sep = c(1,2,3),
           remove = FALSE) %>%
  ## change the col contents from just letters to what they indicate
  mutate(wet_lab = if_else(wet_lab == "N", "NIST", "StJude"),
         seq_lab = if_else(seq_lab == "N", "NIST", "StJude"),
         bioinf = if_else(bioinf == "N", "NIST", "StJude"))

changeseq_anno_df
```

__NOTE__ - The lab, target_site, and/ or changeseq_run variables can be used as grouping variables to calculate within dataset summarise.

## Summary/ Statistical Analysis
### Calculating Summary Metrics

* Total Read Counts per Sample 
```{r}
total_count_df <- changeseq_anno_df %>% 
    group_by(changeseq_run, lab, target_site) %>% 
    summarise(total_read_count = sum(Nuclease_Read_Count))

total_count_df
```

* Total on-target read count
```{r}
target_count_df <- changeseq_anno_df %>% 
    filter(Site_Substitution_Number == 0) %>% 
    rename(on_target_count = Nuclease_Read_Count) %>% 
    select(changeseq_run, lab, target_site, on_target_count)

target_count_df
```


* Percent on-target read count (on-target read count/total read count)
```{r}
on_off_target_df <- left_join(total_count_df, target_count_df) %>% 
    mutate(on_target_rate = on_target_count/total_read_count)

on_off_target_df
```

* Total off-target read count
* Percent off-target read count (off-target read count/total read count)
```{r}
summary_df <- changeseq_anno_df %>% 
    ## replace NAs in Site_Substitution_Number col
    mutate(Site_Substitution_Number = if_else(is.na(Site_Substitution_Number), -1, Site_Substitution_Number), 
           ## assign read count to on-target col of on-target row per sample
            on_target = if_else(Site_Substitution_Number == 0, Nuclease_Read_Count, 0),
           ## assign off-target read counts in non-on-target rows
            off_target = if_else(on_target == 0, Nuclease_Read_Count, 0),
            ## seqs with subs only = seq in Site_Sequence col only
            Substitution_Only = if_else(is.na(Site_Sequence), FALSE, TRUE),
            ## seqs with indels only = seq in Site_Sequence_Gaps_Allowed col only
            Indel_Only = if_else(is.na(Site_Sequence_Gaps_Allowed), TRUE, FALSE),
            ## seqs with two possible reps = seq in both cols
            Sub_or_Indel = if_else(is.na(Substitution_Only || Indel_Only), FALSE, TRUE))
```
  
* Number of sequences with substitutions only
* Number sequences with indels only 
* Number of sequences with two possible representations (substitutions or indels)
```{r}
## create df to plot off-target ratio
off_target_ratio <- summary_df %>%  
    group_by(changeseq_run, lab, target_site) %>% 
    summarise(total_count = sum(Nuclease_Read_Count),
              total_on_target = sum(on_target),
              total_off_target = sum(off_target),
              off_target_ratio = total_off_target/total_count)
## create df to plot alignment of off-target seq resolution details
off_target_seq_align_df <- summary_df%>%    
    group_by(changeseq_run, lab, target_site) %>%
    summarise(total_substitution_only_seqs = sum(Substitution_Only),
              total_indels_only = sum(Indel_Only),
              total_sub_or_indel = sum(Sub_or_Indel))
```

```{r}
off_target_ratio
```

```{r}
off_target_seq_align_df
```


CONSIDERING SEQUENCES THAT HAVE SUBSTITUTIONS ONLY...
* On-target to top 25% reads (on-target read count/top 25% off-target read count)
```{r}
## create df of all off-target rows sorted by decreasing read count value
off_target_df <- summary_df %>%
    mutate(Substitution_Only = if_else(is.na(Site_Sequence), FALSE, TRUE),
           on_target = if_else(Site_Substitution_Number == 0, Nuclease_Read_Count, 0)) %>% 
    filter(Substitution_Only == TRUE) %>%
        arrange(-Nuclease_Read_Count)

## take the top 25% of the off-target df - based on read count value
top_quarter_off_target_df <- off_target_df %>%
    ## Use top_frac to get the top 25% - use group_by to capture within run or site analysis
      top_frac(0.25, wt = Nuclease_Read_Count) %>%
  group_by(changeseq_run, lab, target_site) %>% 
  summarise(top_quarter_offtarget_readcount = sum(Nuclease_Read_Count))
```

```{r}
off_target_df
```

```{r}
top_quarter_off_target_df
```


* On-target to most prevalent off-target read (on-target read count/off-target read with largest read count)
```{r}
## create df to plot ratio between on-target read count and most prevalent off-target read count per sample
on_top_off_df <- summary_df %>% 
    mutate(Site_Substitution_Number = if_else(is.na(Site_Substitution_Number), -1, Site_Substitution_Number), 
            on_target = if_else(Site_Substitution_Number == 0, Nuclease_Read_Count, 0),
           off_target = if_else(on_target == 0, Nuclease_Read_Count, 0)) %>% 
    group_by(changeseq_run, lab, target_site) %>% 
    summarise(on_target_count = max(on_target),
              off_target_max = max(off_target)) %>% 
    mutate(on_off_max_ratio = on_target_count/ off_target_max)

on_top_off_df
```

* Read count for seqs w/ 1 substitution from target sequence
* Read count for seqs w/ 2 substitutions from target sequence
* Read count for seqs w/ 3 substitutions from target sequence
* Read count for seqs w/ 4 substitutions from target sequence
* Read count for seqs w/ 5 substitutions from target sequence
* Read count for seqs w/ 6 substitutions from target sequence
```{r}
## create df to plot the binned read counts of off-target substitution numbers
substitution_count_df <- summary_df %>%
    mutate(Substitution_Only = if_else(is.na(Site_Sequence), FALSE, TRUE)) %>% 
    filter(Substitution_Only == TRUE,
           Site_Substitution_Number > 0) %>% 
    group_by(changeseq_run, lab, target_site, Site_Substitution_Number) %>% 
    summarise(substitution_read_count = sum(Nuclease_Read_Count))

substitution_count_df      
```


### Summary Metric Analysis
#### Plots

* Bar Plot of Total Read Counts per Sample
```{r}
ggplot(total_count_df) + 
    geom_bar(aes(x = target_site, y = total_read_count, fill = lab), 
             position = "dodge", stat = "identity") +
    theme_bw() + 
    theme(axis.text.x = element_text(angle = 90)) + 
    labs(x = "Total Read Count", y = "Editing Target Site", 
         fill = "Lab")
```

Sample F for NSS has a drastically lower read count. 
This was found to be a copy/paste error in their manifest file. 
They were pointing to sample E's read fastq files. 
<!-- Do you mean when comparing NNN and NSN? 
Be careful with making statements that the results are identical or does not introduce any variability. 
Think about why they are identical. 
What are the potential sources of variability?
Is it expected that they are identical? Could it be a bug in your code? I know we have talked about this but these are important questions to ask of future datasets.
-->
For all other samples, my bioinformatics run of CHANGE-Seq is identical to their results in total read counts across the target sites which means that the person running the bioinformatics pipeline doesn't introduce variability - in this metric, at least. 
Regarding the sequencing at NIST vs. the sequencing at St. Jude, our sequencing runs have consistently lower total read counts. 





* Bar Plot of On-Target Read Counts per Sample
```{r}
ggplot(target_count_df) + 
    geom_bar(aes(x = target_site, y = on_target_count, fill = lab), 
             position = "dodge", stat = "identity") +
    theme_bw() + 
    theme(axis.text.x = element_text(angle = 90)) + 
    labs(x = "Total On-Target Read Count", y = "Editing Target Site", 
         fill = "Lab")
```
Consistent pattern (disregarding sample F from St. Jude's sequenced and bioinformatics) of identical values between my run and St. Jude's run of CHANGE-Seq on their sequenced samples and NIST with lower values for reads except for sample C and D where our on-target read count is just slightly more than that of St. Jude's sequenced samples.  


*Scatter Plot of On-Target Read Count/Total Read Count
``````{r}
ggplot(on_off_target_df) + 
    geom_point(aes(x = target_site, y = on_target_rate, fill = lab), 
               shape = 21, alpha = 0.5) + 
    theme_bw() +
    theme(axis.text.x = element_text(angle = 90)) +
        labs(x = "Editing Target Site", y = "On Target Rate", 
         fill = "Lab")
```
On-target rates for sample A, E, and F are nearly matched and variance for samples B, C, and D is observed. NIST-sequenced samples are slightly higher.

*Scatter Plot of Off-Target Read Count/Total Read Count
``````{r}
ggplot(off_target_ratio) + 
    geom_point(aes(x = target_site, y = off_target_ratio, fill = lab), 
               shape = 21, alpha = 0.5) + 
    theme_bw() +
    theme(axis.text.x = element_text(angle = 90)) +
        labs(x = "Editing Target Site", y = "Off Target Rate", 
         fill = "Lab")
```
The rates of the off-target sequences are generally all higher than that of the on-target rates. Like the on-target rates, samples A, E, and F are similar in value. Samples B, C, and D have variance with the inverse of the above - the off-target rates are higher for the St. Jude-sequenced samples than that of the NIST-sequenced samples. 


* Paneled bar plot showing off-target sequence alignment details
```{r}
off_target_seq_align_df %>% 
    ## First make the table long
    gather(key = "metric", value = "value", -changeseq_run, -lab, -target_site) %>% 
    ggplot() +
    geom_bar(aes(x = metric, 
                 y = value,
                 fill = lab),
             position = "dodge", 
             stat = "identity"
             ) +
    ## Split plot by summary metric
    facet_wrap(~target_site, scales = "free_y") +
    theme_bw() +
    theme(axis.text.x = element_text(angle = -45, hjust = 0))
```
The off-target sequences are aligned to the target sequence. These are binned in the following manner: in order to align properly to the target sequence, the off-target sequence could be resolve with substitutions (and assigned a substitution number), gaps (insertions or deletions) or by both possibilities. Here, I have binned them accordingly. As before, no differences are observed between the samples sequenced at St. Jude but run through the CHANGE-Seq bioinformatics pipeline by me and St. Jude, however, variance is observed between the samples sequenced at NIST vs. St. Jude. For samples A and B, NIST-sequenced off-target reads are overall in lower numbers as compared to St. Jude.


* Bar plot of top 25% off-target read counts
```{r}
ggplot(top_quarter_off_target_df) + 
    geom_bar(aes(x = target_site, y = total_read_count, fill = lab), 
             position = "dodge", stat = "identity") +
    theme_bw() + 
    theme(axis.text.x = element_text(angle = 90)) + 
    labs(x = "Total Read Count of Top 25% Off-Targets", y = "Editing Target Site", 
         fill = "Lab")
```
* Scatter plot showing the ratio of on-target read counts to the off-target sequence with the highest read count
``````{r}
ggplot(on_top_off_df) + 
    geom_point(aes(x = target_site, y = on_off_max_ratio, fill = lab), 
               shape = 21, alpha = 0.5) + 
    theme_bw() +
    theme(axis.text.x = element_text(angle = 90)) +
        labs(x = "Editing Target Site", y = "On-Target Rate to Most Prevalent Off-Target", 
         fill = "Lab")
```
As seen with the on-target rate considering total read counts, NIST has higher rates than St. Jude-sequenced samples. 

* Plot of read counts per substitution number for each sample
```{r}
substitution_count_df %>% 
  ggplot() + 
  geom_point(aes(x = Site_Substitution_Number, y = substitution_read_count, fill = lab), 
             shape = 21, alpha = 0.25) + 
  facet_wrap(~target_site, scales = "free_y") + 
  theme_bw()
```

####################################### EVERYTHING ABOVE TRANSFERRED TO FINAL DOC 2020-08-10

#### Table
Replacing code from `changeseq_stats.R`

_Table structure_  

__NOTE__ Might want to rename columns for consistency, e.g. total_substitution_only_seqs to total_subs_only, and remove redundant columns.

```{r}
## Making substitution count data frame wide for join
sub_counts_wide_df <- substitution_count_df %>% 
  mutate(Site_Substitution_Number = paste("sub",Site_Substitution_Number, "count", 
                                          sep = "_")) %>%
  pivot_wider(names_from = Site_Substitution_Number, 
              values_from = substitution_read_count, values_fill = 0)

## Combining summary metric data frames
summary_table <- left_join(off_target_ratio, off_target_seq_align_df) %>% 
  left_join(top_quarter_off_target_df) %>% 
  left_join(on_top_off_df) %>% 
  left_join(sub_counts_wide_df)

## write out summary table
## Note - do not need to use paste with here, here converts the vector of character strings to a file path
write_tsv(summary_table, path = here("data", "summary_stats_table.tsv"), 
          na = "NA", append = FALSE, col_names = TRUE, quote_escape = "double")
summary_table
```



### Summary Table 
```{r}
## Potentially use the DT or other packages for visually pleasing table presentation
summary_table %>% DT::datatable()
```

**Summary Stats Table Column Headers**

 [1] **"changeseq_run":** lab identifiers, sample letter ID, target site                  
 [2] **"lab":** 3-letter identifiers for wet lab, sequencing lab, and bioinformatics lab; **N**IST, **S**t. Jude                           
 [3] **"target_site":** target site                   
 [4] **"total_count":** total read count for that sample (both on and off-target)                   
 [5] **"total_on_target":** total read count for the on-target sequence for that sample                
 [6] **"total_off_target":** total read count for the off-target sequences for that sample                
 [7] **"off_target_ratio":** total on-target read count/total off-target read count               
 [8] **"total_substitution_only_seqs":** total read count for the off-target sequences that are resolved in alignment with nucleotide substitutions only  
 [9] **"total_indels_only":** total read count for the off-target sequences that are resolved in alignment with gaps (indels) only              
 [10] **"total_sub_or_indel":** total read count for the off-target sequences that are resolved in alignment with substitutions or gaps             
 [11] **"top_quarter_offtarget_readcount":** the total read count for the top quarter of off-target reads (sorted by decreasing read count) 
 [12] **"off_target_max":** the read count for the most prevalent off-target sequence (off-target sequence with largest read count)                 
[13] **"on_off_max_ratio":** on-target read count/read count for most prevalent off-target sequence                 
[14] **"Site_Substitution_Number":** each sample is represented in "changeseq_run" with all unique number of substitutions found in that sample - reported in this column       
[15] **"substitution_read_count":** the total read count for each substitution number  



## Off-Target Sequence Comparison
re-implementation of code from `CHANGEseq_compare_fulldf.R` and `edited_CHANGEseq_compare_fulldf.R`. 
These scripts aim to reproduce manuscript Figure 1i. 


# Session Information
## System Information
```{r}
sessioninfo::platform_info()
```


## Package Versions
```{r}
sessioninfo::package_info() %>% 
    filter(attached = TRUE) %>% 
    select(package, loadedversion, date, source) %>%
    knitr::kable(booktabs = TRUE, row.names = FALSE)
```